controller/controller_test.go

```diff
@@ -192,3 +194,104 @@ func Test_validateUnitStatusWithRequest(t *testing.T) {
 		}
 	}
 }
+
+func Test_matchesGroupSlices(t *testing.T) {
+	testCases := []struct {
+		InputUnitName string
+		InputRequest  Request
+		Output        bool
+	}{
+		{
+			"demo-main@1.service",
+			Request{
+				Group:    "demo",
+				SliceIDs: []string{"1", "2"},
+			},
+			true,
+		},
+		{
+			"demo-main@1.service",
+			Request{
+				Group:    "demo",
+				SliceIDs: []string{"3"},
+			},
+			false,
+		},
+
+		{
+			"demo-main.service",
+			Request{
+				Group:    "demo",
+				SliceIDs: nil,
+			},
+			true,
+		},
+
+		{
+			"other-main.service",
+			Request{
+				Group:    "demo",
+				SliceIDs: []string{"1", "2"},
+			},
+			false,
+		},
+		{
+			"other-main.service",
+			Request{
+				Group:    "demo",
+				SliceIDs: nil,
+			},
+			false,
+		},
+	}
+
+	for id, test := range testCases {
+		matcher := matchesGroupSlices(test.InputRequest)
+		result := matcher(test.InputUnitName)
+
+		if result != test.Output {
+			t.Errorf("TestCase %d: Failed to match: Input '%s', expected %v, got %v", id, test.InputUnitName, test.Output, result)
+		}
+	}
+}
+
+// GivenController returns a controller where the fleet backend is replaced
+// with a mock.
+func GivenController() (Controller, *fleetMock) {
+	fleetMock := fleetMock{}
+	cfg := Config{
+		Fleet: &fleetMock,
+	}
+	return NewController(cfg), &fleetMock
+}
+
+func TestController_Start(t *testing.T) {
+	RegisterTestingT(t)
+
+	// Mocks
+	controller, fleetMock := GivenController()
+	fleetMock.On("GetStatusWithMatcher", mock.AnythingOfType("func(string) bool")).Return(
+		[]fleet.UnitStatus{
+			{
+				Name: "test-main@1.service",
+			},
+			{
+				Name: "test-sidekick@1.service",
+			},
+		},
+		nil,
+	).Once()
+	fleetMock.On("Start", "test-main@1.service").Return(nil).Once()
+	fleetMock.On("Start", "test-sidekick@1.service").Return(nil).Once()
+
+	// Execute test
+	req := Request{
+		Group:    "test",
+		SliceIDs: []string{"1", "2"},
```

Comment:
At first glance this test looks wrong, because you think that this cannot work due to the missing units in the request. I guess some mocking magic makes that work somehow.